lib/propolis/src/firmware/smbios/table.rs

```diff
@@ -184,6 +711,213 @@ impl Table for Type16 {
     }
 }
 
+pub mod type16 {
+    enum_try_from! {
+        /// Memory array location.
+        ///
+        /// See Table 72 in section 7.17.1 of [the SMBIOS Reference
+        /// Specification][DSP0136] for details.
+        ///
+        /// [DSP0136]:
+        ///     https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.7.0.pdf
+        #[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+        pub enum Location: u8 {
+            /// Other
+            Other = 0x01,
+            /// Unknown
+            #[default]
+            Unknown = 0x02,
+            /// System board or motherboard
+            SystemBoard = 0x03,
+            /// ISA add-on card
+            IsaCard = 0x04,
+            /// EISA add-on card
+            EisaCard = 0x05,
+            /// PCI add-on card
+            PciCard = 0x06,
+            /// MCA add-on card
+            McaCard = 0x07,
+            /// PCMCIA add-on card
+            PcmciaCard = 0x08,
+            /// Proprietary add-on card
+            ProprietaryCard = 0x09,
+            /// NuBus
+            NuBus = 0x0A,
+            /// PC-98/C20 add-on card
+            Pc98C20Card = 0xA0,
+            /// PC-98/C24 add-on card
+            Pc98C24Card = 0xA1,
+            /// PC-98/E  add-on card
+            Pc98ECard = 0xA2,
+            /// PC-98/Local bus add-on card
+            Pc98LocalCard = 0xA3,
+            // CXL add-on card
+            CxlCard = 0xA4,
+        }
+    }
+
+    enum_try_from! {
+        /// Memory array use field.
+        ///
+        /// See Table 73 in section 7.17.2 of [the SMBIOS Reference
+        /// Specification][DSP0136] for details.
+        ///
+        /// [DSP0136]:
+        ///     https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.7.0.pdf
+        #[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+        pub enum ArrayUse: u8 {
+            /// Other
+            Other = 0x1,
+            /// Unknown
+            #[default]
+            Unknown = 0x2,
+            /// System memory
+            System = 0x3,
+            /// Video memory
+            Video = 0x4,
+            /// Flash memory
+            Flash = 0x5,
+            /// Non-volatile RAM
+            NonVolatile = 0x6,
+            /// Cache memory
+            Cache = 0x7,
+        }
+    }
+
+    enum_try_from! {
+        /// Memory array error correction field.
+        ///
+        /// See Table 74 in section 7.17.3 of [the SMBIOS Reference
+        /// Specification][DSP0136] for details.
+        ///
+        /// [DSP0136]:
+        ///     https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.7.0.pdf
```

Comment:
Could probably put this reference at the top somewhere, rather than repeating it at each bitfield.